[WIP] SSSE3/SSE2, NEON, and Scalar ports, minor fixes

easyaspi314 opened the pull request:
Don't merge yet, it is far from ready.

- SSSE3 port added
   - compatible with SSE2 with a provided polyfill for _mm_alignr_epi8
   - 32-bit safe
   - Optimized for register pressure and code size
     - Full version is rerolled, but welcome to be unrolled by the
       compiler.
 - Working on changing __uint64_t size to size_t
   - Minor speedups on 32-bit targets - there is no reason to use
     uint64_t there because all indices will be < 4billion. Therefore
     64-bit arithmetic is avoided
 - Use int in loops instead of char
   - doesn't mask on ARM
 - Making arguments __restrict
 - NEON port added
 - Scalar port added
   - Portable, endian safe, alignment safe, etc
   - Right now uses subloops
 - XOR in SIMD in the startup
   - prevents vinsert spam on Clang
   - need ideas for improving it
 - Working on fixing linker issues due to inline memes
   - the only inline is static inline

The results have not been fully verified or benchmarked yet.

It is very ugly right now, and I am trying to find a way to structure it
better, but it is a start for now.

espadrine commented:
Reading the code got a grin on my face that isn’t going away. This is really, really good work!

I wonder, do you reckon it would be neater to have it all in the same `.h`?
With preprocessor flags to automatically compile to the best implementation available?

easyaspi314 commented:
> Reading the code got a grin on my face that isn’t going away.

You might want to get that checked out 😛

> I wonder, do you reckon it would be neater to have it all in the same .h?

I am thinking have the portable one be shishua.h, and the others be separate. That allows different versions to easily be recompiled with different flags to dispatch.

As for preprocessor flags:

AVX2 on x86_64:
```c
#if (defined (__x86_64__) || defined(_M_X64)) && defined(__AVX2__)
```
SSE2/SSSE3:
```c
#if defined(__SSE2__) || defined(_M_X64) || defined(__x86_64__) || (defined(_M_IX86_FP) && _M_IX86_FP >= 2)
```
NEON:
```c
#if defined(__ARM_NEON) || defined(__ARM_NEON__)
```

espadrine commented:
That sounds very good!

easyaspi314 commented:
Um, question, you switched `STEPS` and `ROUNDS` between `shishua.h` and `shishua-half.h`. Which is correct?
This drove me nuts trying to debug.

`shishua.h`:
```c
for (char i = 0; i < STEPS; i++) {
    prng_gen(&s, buf, 16 * ROUNDS);
    s.state[0] = s.output[3]; s.state[1] = s.output[2];
    s.state[2] = s.output[1]; s.state[3] = s.output[0];
  }
```
`shishua-half.h`:
```c
for (char i = 0; i < ROUNDS; i++) {
    prng_gen(&s, buf, 4 * STEPS);
    s.state[0] = s.state[1];
    s.state[1] = s.output;
  }
```

Also, is there any reason you are using `__uint64_t` over `uint64_t`? Unlike `__uint128_t`, these are not predefined by the compiler, and instead it bets on it getting defined somewhere from libc in the other headers you include. It is much safer to explicitly include `stdint.h` and use the typedefs defined in the C standard.

espadrine commented:
> Um, question, you switched STEPS and ROUNDS between shishua.h and shishua-half.h. Which is correct?

Naming is hard! The SHISHUA-half naming is a bit more appropriate, though; I forgot to switch `shishua.h` to it.

> Also, is there any reason you are using __uint64_t over uint64_t?

Laziness? Not a good reason at least. Don’t hesitate to do the safer thing and include `stdint.h`.

easyaspi314 commented:
Going to open a new PR to fix some initial issues I noticed first, and rebase from there.

easyaspi314 commented:
As for other things, I found what appears to be a cleaner and smaller path for scalar, and fixed a few bugs.

Here is what I have for `shishua-half-scalar.h` (haven't updated the comments yet, so I left them out for now):

It compiles smaller and doesn't seem to affect unrolling, at least not in Clang.
```c
static inline void prng_gen(prng_state *state, uint64_t *buf, size_t size) {
  for (size_t i = 0; i < size; i += 4) {
    uint64_t t[8];

    memcpy(&buf[i], state->output, sizeof(state->output));

    for (size_t j = 0; j < 4; j++) {
      state->state[j + 4] += state->counter[j];
      state->counter[j] += 7 - (j * 2); // 7, 5, 3, 1
    }
    const uint8_t rotates[16] = { 2,3,0,1, 5,6,7,4,   // left
                                  3,0,1,2, 6,7,4,5 }; // right
    for (size_t j = 0; j < 8; j++) {
      t[j] = (state->state[rotates[j]] >> 32) | (state->state[rotates[j + 8]] << 32);
    }

    for (size_t j = 0; j < 4; j++) {
      uint64_t u_lo = state->state[j + 0] >> 1;
      uint64_t u_hi = state->state[j + 4] >> 3;

      state->state[j + 0] = u_lo + t[j + 0];
      state->state[j + 4] = u_hi + t[j + 4];

      state->output[j] = u_lo ^ t[j + 4];
    }
  }
}
```
Amusingly, the best way to do a scalar `shishua` is to do the shuffle first (`shushia`?) as it avoids an entire temp array.

easyaspi314 committed:
Rename shishua.h and shishua-half.h to *-avx2.h

Preparation for new setup

espadrine commented:
Hi! I wanted to tweak the common `prng_gen()` interface to make it easier to reason about.
It is a small change: it used to take a 64-bit buffer, and now it takes a byte buffer.

Turns out that confusion caused a mistake in the ChaCha8 code which is resolved.
No more seed correlations!

easyaspi314 commented:
> It is a small change: it used to take a 64-bit buffer, and now it takes a byte buffer.

I also have some questions:

 1. Since we are using bytes, should I byteswap on big endian in the scalar version? It makes things more complicated, but ensures that things are reproducible. Note that the internal state itself is endian-safe, and big endian machines will just get the same numbers in reverse order.
 2. How should we handle uneven chunks? Error, UB, or partial writes?

espadrine commented:
> Since we are using bytes, should I byteswap on big endian in the scalar version?

Good question. I think we should ensure that the fingerprint stays the same.

>  should we handle uneven chunks? Error, UB, or partial writes?

For now, I assumed that the buffers given to each PRNG were sized as a multiple of the given number of bytes in the comment. In the future, we can have a buffer.

easyaspi314 committed:
Merge, add scalar implementation

 - Merge buffer type change from master.
 - Add improved scalar implementations
   - Automatically includes SIMD headers if detected, but can be used
     standalone by setting `-DSHISHUA_TARGET=SHISHUA_TARGET_SCALAR`
     - Similarly, the SIMD headers can also be used standalone.
   - Endian safe
   - More simple and compact than previous implementation
Minor changes:
 - Accepts a null buf (in which it will ignore the output)
   - null buf will just advance the state
   - removes buffer from prng_init
 - Asserts on a buffer size that is not a multiple of 32/128 until we
   come up with a proper solution
 - Can now do `gcc prng.c -DHEADER='"shishua.h"'` instead of symlinking

NEON and SSE will be implemented soon.

easyaspi314 commented:
Redoing this to make it smoother.

shishua.h is now shishua-avx2.h, and shishua.h now contains the scalar implementation which automatically includes the correct SIMD header if not told otherwise. This means you can drop `shishua.h` alone and define `-DSHISHUA_TARGET=SHISHUA_TARGET_SCALAR` to just get the basic implementation by itself.


Will add the other implementations soon, I need to really tidy things up as it is a sloppy mess at the moment.

I also made it so a null buf will only advance the state, assert on an uneven output buf, and made it so you can define `-DHEADER='"header.h"'` when compiling prng.c instead of having to symlink.

easyaspi314 commented:
Any feedback on this before I continue working on the SSE and NEON variants?

easyaspi314 committed:
Implement SSE2 and NEON ports

easyaspi314 committed:
Add test vectors

 - Verifies 512 bytes of shishua and shishua half
 - Checks unseeded (seed == 0) and seeded (seed == pi in big endian)
 - Provided program used to generate it in case the algorithm changes

Big endian (MIPS64 via QEMU) has been tested, by the way, and it passes.

easyaspi314 commented:
I implemented SSE and NEON, and also added test vectors so it is easy to verify.

easyaspi314 committed:
Speed tweaks for NEON and Scalar

 - Use pointer arithmetic instead of indexing
 - Don't autoselect NEON path on GCC - GCC has terrible NEON codegen
 - Slight tweaks

easyaspi314 committed:
Fix portability issues

easyaspi314 committed:
More portable Makefile/prng benchmark

prng.c:
 - Now falls back to clock_gettime on non-x86 systems
   - Supports nanosecond accuracy (technically depends on clock resolution)
   - Doesn't need kernel level privledges on AArch64 (unlike
     `__builtin_readcyclecounter`)
   - Not perfect, but portable to most POSIX systems
     - Excludes older iOS versions
 - Accounts for timer latency
   - Tested once with _rdtsc (since it is cycle-based)
   - Checked every time for clock_gettime because it is dependent on
     CPU frequency, slightly slows down performance though.
 - Error on negative/zero buffer size
   - size == 0 is a SIGFPE
 - Added -q/--quiet option
   - Doesn't spit out tons of data to stdout if you don't need it

Makefile:
   - Automatically chooses flags and headers
     - Uses pattern matching to avoid copy-paste issues
     - No longer symlinks
   - Add test vector check
     - change this to the make test instead of doing the whole PractRand
       test suite?

easyaspi314 commented:
Makefile now supports multiple targets and the test vector, prng.c uses `clock_gettime` on non-x86 targets, and I fixed some extra portability issues.

GCC does not benefit from NEON (GCC's NEON is garbage), so I auto select the scalar path and make it opt-in

espadrine commented on shishua.h:
```diff
@@ -1,76 +1,200 @@
+//   shishua half neon   | 0.5169 ns/byte | 0.1929 ns/byte
+//   shishua half scalar | 0.2496 ns/byte | 0.2911 ns/byte
+// Therefore, we only autoselect the NEON path on Clang, at least until GCC's
+// NEON codegen improves.
```
That is a fascinating result!

Thanks for testing this, and smart choice here.

espadrine commented on shishua-avx2.h:
```diff
@@ -0,0 +1,123 @@
+      _mm256_storeu_si256((__m256i*)&buf[i + 64], o2);
+      _mm256_storeu_si256((__m256i*)&buf[i + 96], o3);
+    }
+>>>>>>> Stashed changes
```
I think we want the bottom part, right?

espadrine commented on prng.c:
```diff
@@ -3,27 +3,97 @@
+}
+#else
+// fall back to nanoseconds, e.g. on aarch64 where __builtin_readcyclecounter
+// needs special privledges not usually granted by the Linux kernel.
```
priviledges

espadrine commented on prng.c:
```diff
@@ -38,9 +108,13 @@ args_t parseArgs(int argc, char **argv) {
       a.bytes = strtoll(argv[++i], NULL, 0);
+      if (a.bytes <= 0) a.rval = -1;
+    } else if (strcmp(argv[i], "-q") == 0 || strcmp(argv[i], "--quiet") == 0) {
+      a.quiet = 1;
```
:+1:

espadrine commented on prng.c:
```diff
@@ -3,27 +3,97 @@
+static inline int64_t timer_elapsed(prng_timer_t start) {
+  int64_t end = _rdtsc();
+  // While latency isn't as much of a problem with _rdtsc as it is with
+  // clock_gettime, measure it anyways to be safe.
```
I'm a bit confused by this.

Unlike RDTSCP, RDTSC is not serializing, so given the superscalar part of the CPU pipeline, the latency we measure here will just be parallel noise, no?

Do you get a better measurement based on an error metric?

easyaspi314 replied:
I honestly haven't fully tested this. I just assumed that there would be a little latency. I'll remove it as that probably affects the benchmark.

espadrine commented on prng.c:
```diff
@@ -3,27 +3,97 @@
+#include <time.h>
+
+typedef struct timespec prng_timer_t;
+static const char *unit = "ns/byte";
```
:+1: on having a fallback, nice!

And the unit you picked makes the most intuitive sense comparatively to the cpb metric.

espadrine commented on prng.c:
```diff
@@ -3,27 +3,97 @@
+// needs special privledges not usually granted by the Linux kernel.
+// Note that this depends on the CPU frequency.
+// This assumes that clock_gettime is available.
+#include <time.h>
```
Should this be indented, as it is in the else?

espadrine commented on Makefile:
```diff
@@ -1,57 +1,52 @@
+##
+
+# Replace pseudo target names with the real names
+fix_target = $(subst plus,+,$(subst -scalar,,$(subst -ssse3,-sse2,$(1))))
```
Should we just convert all the `+` to `plus`? Maybe it's not worth the trouble.

Besides, it accidentally creates two programs for the same thing right now I believe.

espadrine commented on shishua.h:
```diff
@@ -1,76 +1,200 @@
+#  if defined(__AVX2__) && (defined(__x86_64__) || defined(_M_X64))
+#    define SHISHUA_TARGET SHISHUA_TARGET_AVX2
+#  elif defined(__x86_64__) || defined(_M_X64) || defined(__SSE2__) || (defined(_M_IX86_FP) && _M_IX86_FP >= 2)
+#    define SHISHUA_TARGET SHISHUA_TARGET_SSE2
```
Love it, very precise!

I am slightly ignorant on what `_M_X64` is (is it a Windows equivalent to `__x86_64__`?)

easyaspi314 replied:
Yes.

espadrine commented on shishua.h:
```diff
@@ -1,76 +1,200 @@
+      // Lookup table for the _offsets_ in the shuffle. Even lanes rotate
+      // by 5, odd lanes rotate by 3.
+      // If it were by 32-bit lanes, it would be
+      // { 5,6,7,0,1,2,3,4, 11,12,13,14,15,8,9,10 }
```
:+1: on the exhaustive comments!

espadrine commented on shishua-neon.h:
```diff
@@ -0,0 +1,182 @@
+  uint64x2_t counter[2];
+} prng_state;
+
+#if defined(__GNUC__) //&& (defined(__BYTE_ORDER__) && __BYTE_ORDER__==__ORDER_LITTLE_ENDIAN__)
```
Forgotten comment?

easyaspi314 replied:
Oof, good catch. 🤦

espadrine commented on Makefile:
```diff
@@ -164,3 +159,7 @@ benchmark-amd: /usr/bin/gcloud
 	gcloud compute instances delete shishua-amd
 
 .PHONY: test benchmark-intel benchmark-amd
```
We can combine the phonies.

espadrine commented on Makefile:
```diff
@@ -1,57 +1,52 @@
-PRNGS = shishua shishua-half chacha8 xoshiro256plusx8 xoshiro256plus romu wyrand lehmer128 rc4
+CFLAGS := -O3 -g
+FINGERPRINT := $(shell ./shishua -b 256 | ./fingerprint.sh)
+TARGETS := scalar sse2 ssse3 avx2 neon
```
The awkward part of this is that dependencies will break when a target fails.

For instance, a computer which doesn't have arm_neon.h will try `make test/benchmark-perf` (which depends on `$(PRNGS)`), and the make command will fail because of the ARM error.

We can get around that by defining two semantically distinct variables:

- PRNGS is the list of designs (regardless of the implementation): shishua, shishua-half, chacha8, etc.
- IMPLS is the list of implementations (shishua, shishua-neon, etc).

What do you think?

espadrine commented on xoshiro256+x8.h:
```diff
@@ -17,13 +19,32 @@ typedef struct prng_state {
-  for (size_t i = 0; i < n; i += XOSHIRO256_UNROLL) {
-    for (size_t j = 0; j < XOSHIRO256_UNROLL; j++) { buf[i + j] = s->state[0][j] + s->state[3][j]; }
+  for (size_t i = 0; i < n; i += XOSHIRO256_UNROLL * 8) {
+    for (size_t j = 0; j < XOSHIRO256_UNROLL; j++) { prng_write_le64(&buf[i + (8 * j)], s->state[0][j] + s->state[3][j]); }
```
I’m not sure why, but the performance of this is severely degraded, much much slower than Xoshiro256+, when it should be much much faster.

easyaspi314 replied:
Hmm...perhaps it is compiling as a `vmovdqu` instead of `vmovdqa`.

Will have to check. What compiler and flags are you using?

espadrine replied:
I just did `make test/benchmark-perf`.

I realize that a few important flags were lost. The initial command was:

    gcc -O9 -fdisable-tree-cunrolli -march=native -o xoshiro256plux8 prng.c

The new one:

    cc -O3 -g -DHEADER='"xoshiro256+x8.h"' prng.c -o xoshiro256plusx8

espadrine commented on shishua-half-avx2.h:
```diff
@@ -0,0 +1,112 @@
+    if (buf != NULL) {
+      _mm256_storeu_si256((__m256i*)&buf[i], o);
+    }
+>>>>>>> Stashed changes
```
Forgotten merge conflict I think.

easyaspi314 replied:
Yes, I was a little overtired when I committed this. I will fix it soon.

espadrine commented on prng.c:
```diff
@@ -3,27 +3,97 @@
     prng_gen(&s, buf, sizeof(buf));
-    cycles += _rdtsc() - start;
-    ssize_t w = write(STDOUT_FILENO, buf, wbytes);
+    cycles += timer_elapsed(start);
```
The cpb have changed a bit, but I'm not 100% sure why.

easyaspi314 replied:
Is it faster or slower?

erthink replied:
Take look to [this comment](https://github.com/rurban/smhasher/issues/125#issuecomment-625776324), it may be useful for measuring CPU clock cycles.

espadrine replied:
With the current code, it is less precise. We used to have two digits of precision (so, very reproducible results), and now it is ±0.05 cpb. Also, SHISHUA and xoshiro256+x8 is slower, ChaCha8 and a few others are faster.

Bulat-Ziganshin replied:
@erthink when measuring bytes/cycle in code involving AVX, both rdtscp and my approach will fail - rdtscp in intel cpus measures ticks of basic frequency (f.e. it's ~2 GHz for i9-7820 and i7-8665), while my approach employing only scalar ops indeed measures frequency in AVX-less code, while AVX frequency may be lower

Bulat-Ziganshin replied:
Proper use of rdtscp described in https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/ia-32-ia-64-benchmark-code-execution-paper.pdf

(in short: rdtscp + code + rdtscp + save eax/adx/... + cpuid), cpuid ensures that later commands will not started executing before cpuid so rdtscp will count only time necessary to execute code between two rdtscp commands)

easyaspi314 replied:
Hmm, from what I am reading, we don't need to do cpuid, only mfence and lfence, or even better, use rdtscp and lfence if we can.
We should be able to do this:
```c
static inline int64_t timer_start(void)
{
#ifdef __GNUC__
    size_t eax, edx; // avoid manual zero extending on 64-bit
    __asm__ __volatile__(
        "mfence\n"
        "lfence\n"
        "rdtsc"
      : "=a" (eax), "=d" (edx) :: "memory");
    return eax | ((int64_t)edx << 32);
#else
    _mm_mfence();
    _mm_lfence();
    return _rdtsc();
#endif
}

static inline int64_t timer_elapsed(int64_t start)
{
#if defined (__SSE4_2__) || defined(__AVX__)
    // rdtscp is only available on Nehalem and later.
#  ifdef __GNUC__
    size_t eax, edx;
    __asm__ __volatile__(
        "rdtscp\n"
        "lfence"
        : "=a" (eax), "=d" (edx) :: "%ecx", "memory");
    int64_t end = eax | ((int64_t)edx << 32);
#  else
    unsigned dummy;
    int64_t end = _rdtscp(&dummy);
    _mm_lfence();
#  endif
    return end - start;
#else 
    return timer_start() - start;
#endif
}

```

Bulat-Ziganshin replied:
mfence/lfence is MEMORY OPERATION fences while rdtscp+cpuid makes fences around ALL COMMANDS executed by the cpu. this ensures that you measure only operations executed between rdtscp commands.

espadrine reviewed (changes requested):
Looks good! I like the direction this is taking!

easyaspi314 committed:
Fix merge conflicts

Never git commit without sufficient coffee.

easyaspi314 committed:
Fix testing comment

easyaspi314 committed:
Remove latency testing from rdtsc benchmark

Apparently, it should be accurate enough as-is.

espadrine commented:
Is it OK if I merge the changes that don’t affect timings first?

That way we can more easily study what impact the timing changes have, why, and what is the best approach.

easyaspi314 commented:
Yeah, go ahead.

espadrine commented:
I merged this patch!

The few modifications I made are in 7f15b57b43afc773efe1154be92cff4bc088052b
